src/consensus/cfund.cpp

```diff
@@ -384,18 +384,47 @@ bool CFund::CProposal::IsRejected() const {
            && ((float)nVotesNo > ((float)(nTotalVotes) * Params().GetConsensus().nVotesRejectProposal));
 }
 
+bool CFund::CProposal::CanVote() const {
+    return (fState == NIL) && (nVotingCycle <= Params().GetConsensus().nCyclesProposalVoting);
+}
+
 bool CFund::CProposal::IsExpired(uint32_t currentTime) const {
     if(nVersion >= 2) {
         if (fState == ACCEPTED && mapBlockIndex.count(blockhash) > 0) {
             CBlockIndex* pblockindex = mapBlockIndex[blockhash];
             return (pblockindex->GetBlockTime() + nDeadline < currentTime);
         }
-        return (fState == EXPIRED) || (fState == PENDING_VOTING_PREQ) || (nVotingCycle > Params().GetConsensus().nCyclesProposalVoting && (CanVote() || fState == EXPIRED));
+        return (fState == EXPIRED) || (fState == PENDING_VOTING_PREQ) || (nVotingCycle > Params().GetConsensus().nCyclesProposalVoting && (fState == NIL || fState == EXPIRED));
     } else {
         return (nDeadline < currentTime);
     }
 }
 
+std::string CFund::CProposal::GetState(uint32_t currentTime) const {
+    std::string sFlags = "pending";
+    if(IsAccepted()) {
+        sFlags = "accepted";
+        if(fState == PENDING_FUNDS)
+            sFlags += " waiting for enough coins in fund";
+        else if(fState != ACCEPTED)
+            sFlags += " waiting for end of voting period";
+    }
+    if(IsRejected()) {
+        sFlags = "rejected";
+        if(fState != REJECTED)
+            sFlags += " waiting for end of voting period";
+    }
+    if(currentTime > 0 && IsExpired(currentTime)) {
+        sFlags = "expired";
+        if(fState != EXPIRED && nVotingCycle <= Params().GetConsensus().nCyclesProposalVoting)
```

Comment:
We use this check `nVotingCycle <= Params().GetConsensus().nCyclesProposalVoting` all over the place, but it's not really clear as to what it's purpose is, if a proposal is `EXPIRED` why do we need to check the voting cycle is? I understand it's because we don't update the `fState` until the end of the 7th cycle, however to make this clearer I think we should do one of two things:
- Add comments in the code around these checks so it's clear what their purpose is (easiest method but kind of annoying, not keen on this)
- Refactor this check out into a function that has a descriptive name and potentially a single comment in the function that describes it's purpose. 
|e.g. ```
checkIfVotingCycleCapExceeded() {
// this function checks if a proposal who's fState has not yet been set to expired ....etc
}
```